e2e/testcases/hpa_test.go

```diff
@@ -0,0 +1,117 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	appsv1 "k8s.io/api/apps/v1"
+	autoscalingv2 "k8s.io/api/autoscaling/v2"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apiregistrationv1 "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1"
+	"kpt.dev/configsync/e2e/nomostest"
+	"kpt.dev/configsync/e2e/nomostest/ntopts"
+	nomostesting "kpt.dev/configsync/e2e/nomostest/testing"
+	"kpt.dev/configsync/e2e/nomostest/testpredicates"
+	"kpt.dev/configsync/e2e/nomostest/testwatcher"
+	"kpt.dev/configsync/pkg/kinds"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func TestHPA(t *testing.T) {
+	nt := nomostest.New(t, nomostesting.Reconciliation2,
+		ntopts.SyncWithGitSource(nomostest.DefaultRootSyncID, ntopts.Unstructured))
+	rootSyncGitRepo := nt.SyncSourceGitReadWriteRepository(nomostest.DefaultRootSyncID)
+
+	nt.T.Log("Checking if metrics-server is installed")
+	// GKE installs metrics-server in the kube-system namespace,
+	// but our test fixture installs into the metrics-system namespace,
+	// so check both.
+	metricsServerList := &appsv1.DeploymentList{}
+	nt.Must(nt.KubeClient.List(metricsServerList,
+		client.MatchingFields{"metadata.name": "metrics-server"}))
+
+	switch len(metricsServerList.Items) {
+	case 0:
+		// Install if not found.
+		// Normal test cleanup will handle uninstall.
+		nt.T.Log("Deploy metrics-server Deployment & APIService")
+		nt.Must(rootSyncGitRepo.Copy(fmt.Sprintf("%s/metrics-server/components-v0.7.2.yaml", yamlDir), "acme/namespaces/metrics-system/components-v0.7.2.yaml"))
+		nt.Must(rootSyncGitRepo.CommitAndPush("Add metrics-server"))
+		nt.Must(nt.WatchForAllSyncs())
+
+		nt.T.Log("Wait for metrics-server Deployment to be available")
+		nt.Must(nt.Watcher.WatchForCurrentStatus(kinds.Deployment(), "metrics-server", "metrics-system"))
+	case 1:
+		metricsServerNS := metricsServerList.Items[0].Namespace
+		nt.T.Logf("Found metrics-server Deployment in namespace %q", metricsServerNS)
+
+		nt.T.Log("Wait for metrics-server Deployment to be available")
+		nt.Must(nt.Watcher.WatchForCurrentStatus(kinds.Deployment(), "metrics-server", metricsServerNS))
+	default:
+		for _, obj := range metricsServerList.Items {
+			nt.T.Logf("Found metrics-server Deployment in namespace %q", obj.Namespace)
+		}
+		nt.T.Fatal("Too many metrics-server deployments installed")
+	}
+
+	nt.T.Log("Wait for v1beta1.metrics.k8s.io APIService to be available")
+	nt.Must(nt.Watcher.WatchObject(kinds.APIService(), "v1beta1.metrics.k8s.io", "",
+		testwatcher.WatchPredicates(testpredicates.HasConditionStatus(nt.Scheme,
+			string(apiregistrationv1.Available), corev1.ConditionTrue))))
+
+	nt.T.Log("Deploy hello-world Deployment")
+	nt.Must(rootSyncGitRepo.Copy(fmt.Sprintf("%s/hpa/ns-helloworld.yaml", yamlDir), "acme/namespaces/helloworld/ns-helloworld.yaml"))
```

Comment:
minor nit: I prefer to inline small/simple k8s objects like this ns/deployment/hpa so that the test is readable/debuggable without needing to reference external yaml files

Reply:
:shrug: we already have this hello-app used in a few places. So I just left it consistent with those using testdata files. But I hear you.